### PR TITLE
[misc] LoRA - Skip LoRA kernels when not required

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
In V0 LoRA, we skipped the LoRA kernel launch when all the scheduled requests target the base model.
PR https://github.com/vllm-project/vllm/pull/14685 removed this optimization.
This PR re-introduces this optimizations such that it works for both V0 and V1.

Previously, i.e. before #14685 , on V0, we had a boolean flag, `no_lora`,  in `punica_gpu.py` that tracked if any of the input requests needed LoRA. This works well for V0. But the V1 architecture uses the traced torch.compile graphs to execute a forward pass. This tracing, doesn't play well with dynamic control flow. However, the tracing treats `torch.ops` functions as a black box.

This PR, moves the flag `no_lora` flag inside the `lora_expand` and `lora_shrink` torch operations and triggers an early exit from the operation. 

Benchmarks:
server command : 
```
 VLLM_USE_V1=0  vllm serve meta-llama/Llama-2-7b-chat-hf  --gpu-memory-utilization 0.95  --enable-lora --max-loras 3 --max-cpu-loras 15 --max-lora-rank 64 --lora-modules lora=xtuner/Llama-2-7b-qlora-moss-003-sft --port 9001  --no-enable-prefix-caching
```

client command : 
```
python3 benchmarks/benchmark_serving.py --model meta-llama/Llama-2-7b-chat-hf --dataset-name random --random-input-len 2048 --random-output-len 512 --request-rate inf --seed ${i} --port 9001 --num-prompts ${num_prompts}
```

Numbers: 
The `benchmark_serving.py` command was run 4 times for every `num_prompts` values. All `mean_ttft_ms` measurements are provided below, 
<google-sheets-html-origin><style type="text/css"></style>
Machine : 1xA100 |   |   |  
-- | -- | -- | --
mean_ttft_ms values |   |   |  
branch \ Num prompts | 1 | 4 | 8
main (400d483e8)  - has #14685 | 148, 170, 170, 176 | 430, 658, 434, 437 | 750, 747, 762, 737
This PR | 155, 165, 162, 170 | 410, 922, 400, 407 | 713, 687, 799, 724

Thanks @jeejeelee for flagging this 🙌

### CODE REVIEW

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py [hunk 1]
```diff
@@ -17,6 +17,17 @@ class LoRAKernelMeta:
...
+    # The V1 architecture uses the traced torch.compile graphs to execute
+    # a forward pass. Things to note about this process,
+    # 1. The tracing infers all python scalar datatype objects into a constant
+    # value.
+    # 2. The tracing cannot handle dynamic control flow. (dynamic control flow
+    # is an experimental feature in pytorch)
+    # 3. The internals of torch.ops functions are not traced.
+    # We disguise the "no_lora" flag as a cpu tensor and leverage point number 3
+    # to early exit from inside the lora_expand / lora_shrink torch operation.
+    no_lora_flag_cpu: torch.Tensor
```
Comment (varun-sundar-rabindranath) [inline vllm/lora/ops/triton_ops/lora_kernel_metadata.py:29]: @bnellnm @ProExpertProg @youkaichao can you take a look when you find some time please. Thanks 🙌

Comment (youkaichao) [reply vllm/lora/ops/triton_ops/lora_kernel_metadata.py:29]: it is a good summarization. we leverage the point number 3 to deal with complicated attention operations.

Comment (varun-sundar-rabindranath) [reply vllm/lora/ops/triton_ops/lora_kernel_metadata.py:29]: For cudagraphs we always capture with LoRA. so there is just 1 path in that case.

Comment (bnellnm) [reply vllm/lora/ops/triton_ops/lora_kernel_metadata.py:29]: lgtm

File: vllm/lora/ops/triton_ops/lora_expand.py [hunk 2]
```diff
@@ -157,11 +158,18 @@ def _lora_expand(
...
         lora_ids (torch.Tensor): LoRA ids to process.
+        no_lora_flag_cpu (torch.Tensor): A CPU tensor of size 1, that indicates
+            if there are any requests that require LoRA.
         offset_start (int, optional): Offset start for output_tensor. 
             Defaults to 0.
         add_inputs (bool, optional): Whether to add the input tensor to the 
             output tensor. Defaults to False.
     """
+
+    if no_lora_flag_cpu.item():
```
Comment (bnellnm) [inline vllm/lora/ops/triton_ops/lora_expand.py:169]: should we assert somewhere that `no_lora_flag` is a scalar?

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py [hunk 3]
```diff
@@ -70,6 +88,11 @@ def prepare_tensors(self, token_lora_mapping: torch.Tensor) -> None:
 
         self._reset()
 
+        self.no_lora_flag_cpu[0] = torch.all(token_lora_mapping == -1)
+        if self.no_lora_flag_cpu.item():
```
Comment (bnellnm) [inline vllm/lora/ops/triton_ops/lora_kernel_metadata.py:92]: Seems like you could use the result of torch.all directly instead of item()?

File: vllm/lora/ops/triton_ops/lora_shrink.py [hunk 4]
```diff
@@ -126,8 +127,15 @@ def _lora_shrink(
             identifies the region in token_indices_sorted_by_lora_ids that
             LoRA lora_ids[i] should process.
         lora_ids (torch.Tensor): LoRA ids to process.
+        no_lora_flag_cpu (torch.Tensor): A CPU tensor of size 1, that indicates
+            if there are any requests that require LoRA.
         scaling (float): Scaling factor.
     """
+
+    if no_lora_flag_cpu.item():
```
Comment (bnellnm) [inline vllm/lora/ops/triton_ops/lora_shrink.py:135]: ditto about item() alternatives.

File: vllm/lora/ops/triton_ops/lora_expand.py (hunk 2 above)
Comment (varun-sundar-rabindranath) [reply vllm/lora/ops/triton_ops/lora_expand.py:169]: Thanks @bnellnm - I'll check it out.

Comment (varun-sundar-rabindranath) [reply vllm/lora/ops/triton_ops/lora_expand.py:169]: It looks like `.item()` is the best option. I verified this by 
```
>>> setup = '''
... import torch
... x = torch.tensor([True], dtype=torch.bool, device="cpu")
... '''
>>> item_timeit = timeit.timeit(stmt="x.item()", setup=setup, number=100000)
>>> all_timeit = timeit.timeit(stmt="torch.all(x)", setup=setup, number=100000)
>>> data_timeit = timeit.timeit(stmt="x.data[0]", setup=setup, number=100000)
>>> item_timeit
0.05852670222520828
>>> all_timeit
0.2363106533885002
>>> data_timeit
0.1912248134613037
```

Comment (varun-sundar-rabindranath) [reply vllm/lora/ops/triton_ops/lora_expand.py:169]: > Should we assert somewhere that no_lora_flag is a scalar?
Yes . I'll make the change 👍

Comment (bnellnm) [reply vllm/lora/ops/triton_ops/lora_expand.py:170]: > It looks like `.item()` is the best option. I verified this by
> 
> ```
> >>> setup = '''
> ... import torch
> ... x = torch.tensor([True], dtype=torch.bool, device="cpu")
> ... '''
> >>> item_timeit = timeit.timeit(stmt="x.item()", setup=setup, number=100000)
> >>> all_timeit = timeit.timeit(stmt="torch.all(x)", setup=setup, number=100000)
> >>> data_timeit = timeit.timeit(stmt="x.data[0]", setup=setup, number=100000)
> >>> item_timeit
> 0.05852670222520828
> >>> all_timeit
> 0.2363106533885002
> >>> data_timeit
> 0.1912248134613037
> ```

I went and found some relevant links about `.item()`:
https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#avoid-unnecessary-cpu-gpu-synchronization
https://discuss.pytorch.org/t/synchronization-slow-down-caused-by-item-which-is-not-caused-by-data-0/19394

Calling it in a tight loop with no dependencies may be fast but it might be bad due to the synchronization issues mentioned in those links.  I don't know of a good way to verify this though.

Comment (bnellnm) [reply vllm/lora/ops/triton_ops/lora_expand.py:170]: > x = torch.tensor([True], dtype=torch.bool, device="cpu")

Is the flag always on the CPU or the GPU?  I guess if it is on the CPU then those guidelines don't apply.

File: vllm/worker/model_runner.py [hunk 5]
```diff
@@ -1479,6 +1479,18 @@ def capture_model(self, kv_caches: List[List[torch.Tensor]]) -> None:
                 dtype=self.model_config.dtype,
```
[the remark this answers is not in the document]
Comment (jeejeelee) [reply vllm/worker/model_runner.py:1479]: Thank you for the detailed explanation, I like this.

File: vllm/worker/model_runner.py [hunk 6]
```diff
@@ -1479,6 +1479,18 @@ def capture_model(self, kv_caches: List[List[torch.Tensor]]) -> None:
...
+        if self.lora_config:
+            assert self.lora_manager is not None
+            dummy_lora_id = 1
+            dummy_lora_request = LoRARequest(
+                lora_name=f"dummy_{dummy_lora_id}",
+                lora_int_id=dummy_lora_id,
+                lora_path='/not/a/real/path')
+            self.lora_manager.add_dummy_lora(dummy_lora_request,
+                                             LORA_WARMUP_RANK)
+
```
Comment (jeejeelee) [inline vllm/worker/model_runner.py:1493]: During the warmup, we added some dummy LoRAs, then removed them. Perhaps we could continue using those dummy LoRAs and remove them after the capture is complete. I think this would reduce redundant code. See : https://github.com/vllm-project/vllm/blob/main/vllm/worker/worker.py#L324

Comment (varun-sundar-rabindranath) [reply vllm/worker/model_runner.py:1503]: The `GPUModelRunnerBase::_dummy_run` is the one adding and removing the dummy LoRAs inside it. This `_dummy_run` function is used in 2 places,
1. GPUModelRunnerBase::profile_run
2. Worker::_warm_up_model
Updating _dummy_run to act differently based on the caller seems cumbersome. 
But, I agree with your point on redundant code. I have refactored adding and removing of the dummy loras in this commit https://github.com/vllm-project/vllm/pull/15152/commits/c05763eb38ad1457b295650175c6b2b9f3c11819 Please take a look. 
Note : I considered using a context manager, but it looks like we already have a lot of indentation in that code and I didn't want to add another level. 
What do you think ?

**Review by jeejeelee — APPROVED** [review]
